samples/spring-eventsourced-shopping-cart/src/main/java/com/example/shoppingcart/domain/ShoppingCartEvent.java

```diff
@@ -4,21 +4,20 @@
 import com.fasterxml.jackson.annotation.JsonSubTypes;
 import com.fasterxml.jackson.annotation.JsonTypeInfo;
 
-@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
-@JsonSubTypes(
+// tag::events[]
+@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type") // <1>
+@JsonSubTypes( // <2>
     {
         @JsonSubTypes.Type(value = ShoppingCartEvent.ItemAdded.class, name = "item-added"),
         @JsonSubTypes.Type(value = ShoppingCartEvent.ItemRemoved.class, name = "item-removed"),
         @JsonSubTypes.Type(value = ShoppingCartEvent.CheckedOut.class, name = "checked-out")
     })
-public sealed interface ShoppingCartEvent {
+public sealed interface ShoppingCartEvent { // <3>
 
-  record ItemAdded(ShoppingCart.LineItem item) implements ShoppingCartEvent {
-  }
+  record ItemAdded(ShoppingCart.LineItem item) implements ShoppingCartEvent { }
```

Comment:
And alternative here is to use `@JsonTypeName` to annotated each record. 

If we use it, we can remove the `name` attribute from the `@JsonSubTypes.Type` above. 

Anyway, that's just a taste thing. Both techniques works.

Reply:
Hmm, now I'm having a second thought about what I just said. 

Having the name declared together with the type in the `@JsonSubTypes.Type` may be a better option. 

My reasoning is that each time that a user adds a new event type, they need to add a new `@JsonSubTypes.Type` line. As such, they have only one place to care about and they also see the other lines, so they will remember that they need to choose a name for the type hint. 

My point is, using `@JsonSubTypes.Type` like we are doing here should be the recommended technique as it increases the chances of doing the right thing.